text/0000-roadmap-2020.md

```diff
@@ -0,0 +1,368 @@
+> are in this state for a long time.
+>
+> -- [vorner](https://vorner.github.io/2019/11/12/rust-2020.html)
+
+A major theme highlighted in numerous blog posts and team member's
+feedback is the tendency for Rust efforts to sometimes "get stuck"
+without being fully completed. Over the years, Rust has accumulated a
+number of "almost complete" efforts -- these range from
+language/library features to compiler refactorings to community
+projects.
+
+One of our goals for this year is to reduce this backlog of "in
+progress" ideas, whether by implementing them or by explicitly opting
+to reject or postpone the idea. This does not mean that we should not
+accept any new work, but we should have a high level goal in mind of
+finishing the year with less, rather than more, "planned" work.
+
+There are several motivations here. First, the set of "in-progress"
+designs and efforts already encompasses the most hotly desired
+features and initiatives. But further, stalled work can be
+demotivating and confusing. When an initiative spans over several
+years, it becomes harder and harder to track the current the state and
+to remember all of the key design constraints. This in turn hinders
+participation in the Rust project and makes it harder to figure out
+what is going on (see also: the goal of improving visibility into the
+state of our initiatives and design efforts).
+
+## Improve project functioning, governance, and visibility
+
+> Organizational work is at the core of everything else that happens in the project, and above all else this seems to be the one thing we should keep improving. We’re growing fast, and our organization needs to grow with it.
+>
+> -- [Yoshua Wuyts](https://blog.yoshuawuyts.com/rust-2020/)
+
+The Rust project has grown dramatically over the last few years, in every dimension:
+
+* We have more users than ever before.
+* We are seeing many more companies -- and much larger companies -- adopting Rust.
+* Our organization and Rust teams have grown.
+
+This is great news! But with this growth comes challenges.  We are
+finding that it is harder and harder to ensure communication across
+the organization. It can often be challenging to find enough people to
+do the work we would like to get done, which in turn leads to burnout
+or people leaving the project. We've identified three major goals that
+we think will help.
+
+### Improve visibility into the state of initiatives and design efforts
+
+Right now it is very difficult to answer questions like "what are the
+active efforts and how can I help" to "what is the status of feature
+X". This is true both for folks who are deeply embedded in the Rust
+organization and for newcomers.
+
+There are many ways to improve visibiility, but the most basic step is
+simply expending more effort on posting updates and documenting the
+status.  Things like the Inside Rust blog are helpful here, and we
+should look for other ways to incorporate lightweight status updates
```

Comment:
I'm pretty happy with the new strategies of a) posting implementation history updates in tracking issues, e.g. as in https://github.com/rust-lang/rust/issues/54883, b) using feature gate labels (e.g. `F-const_generics`).

Reply:
(Resolving -- it seems like this is not a concern? :)

I will say that I agree (and I think we all do) that we have been doing better in this area, but I think it is also true that we could do a lot better too.

Reply:
I thought it was worth mentioning *in the text* as useful information on the progress made thus far. I'm actually not sure we could do so much better other than perhaps some automation.

Reply:
One big audience we aren't serving with those strategies is the person who has never visited the Rust issue tracker. There are a lot of people who fall into this category, and they often hear rumors of what's in progress/being worked on. I think it'd be nice to serve these people in a way better than pointing them at the issue tracker.

Reply:
@Centril I strongly and respectfully disagree.

I'm sure the teams internally do a good job of keeping track of stuff, but unless you are on one of the teams, the process is very opaque. You have to actively lurk on zulip, discord, internals, and github, and read TWiR religiously to find out what is being worked on, what recent developments are, what team priorities are, etc. As a maintainer of rustc-guide and member of WG-learning, I consider myself relatively knowledgeable, and I still feel somewhat left in the dark.

In the ideal case, if one wanted highest visibility to newbies, there would be something like a giant kanban board somewhere that lists what everyone is doing/prioritizing across all teams, what's blocked, what's being worked on, what is de-prioritized, what is urgent, and it would be updated say weekly. Of course, I realize that's a lot of work, but perhaps that's the ideal to work towards.

To be clear, I'm not saying the teams should do this for themselves; whatever process you're already using seems to work fine. I'm saying there should be some medium intended for consumption by non-team-members that accomplishes this purpose. I think the "Inside Rust" blog is a good step in this direction.

Reply:
> You have to actively lurk on zulip, discord, internals, and github, and read TWiR religiously to find out what is being worked on, what recent developments are, what team priorities are, etc.

You can also subscribe to the tracking issues you are interested in and get updates there. (Unless the tracking issue is poorly maintained, which is the case for the old ones, and not everyone is following the new process of feature gate labels and whatnot, so I need to do some PSAs about that). However, the primary function of these labels and tracking issues is to keep track of the bugs, open issues, and history for review by the teams so that everything is covered. I think it does function well for finding 

> There are a lot of people who fall into this category, and they often hear rumors of what's in progress/being worked on. I think it'd be nice to serve these people in a way better than pointing them at the issue tracker.

> In the ideal case, if one wanted highest visibility to newbies, there would be something like a giant kanban board somewhere that lists what everyone is doing/prioritizing across all teams, what's blocked, what's being worked on, what is de-prioritized, what is urgent, and it would be updated say weekly. Of course, I realize that's a lot of work, but perhaps that's the ideal to work towards.

Could you please articulate why it is important that we serve newbies with the highest visibility for unfinished features that are being worked on beyond what e.g. https://github.com/rust-lang/rust/issues/54883 would offer? I would personally find it stressful to report exactly what I'm doing, what I'm prioritizing, etc. on a weekly basis, and it would make me feel more like an employee in a company, developing a commodity, rather than working on FOSS. I think saying that we should work towards this ideal needs to be, but hasn't been justified.

> To be clear, I'm not saying the teams should do this for themselves; whatever process you're already using seems to work fine. I'm saying there should be some medium intended for consumption by non-team-members that accomplishes this purpose. I think the "Inside Rust" blog is a good step in this direction.

As long as we can continue tracking issues like in https://github.com/rust-lang/rust/issues/54883 (as opposed to moving out our tracking to some other place where I cannot use feature gate labels, links and backlinks to issues and PRs) and that this extra "newbie friendly" work doesn't have to be done by language team members then I have no objections.

Reply:
I would argue that the following are highly desirable (and maybe to some extent, crucial):

- It should be easy for contributors (especially new contributors) to see what is being worked on, what priorities are there, etc
- It should be easy for users to see where the language is headed, what features are likely or unlikely to land, etc

By "easy" what I mean is that a person who makes a reasonable effort should be able to gain visibility. IMHO this is not the case, as I mentioned before.

Note that these goals are broader than a single issue; they are project-level, so tracking issues are insufficient to solve this problem. I think this is why there are often questions like "any progress on this?" on tracking issues.

To be clear, I'm not arguing that we should actually make the giant kanban board.  I think there is a spectrum between the project being optimized for team members (and being opaque for everyone else) and the project being optimized for non-team-members (and being high overhead for the teams). I am arguing that perhaps we need adjust where on the spectrum we fall to be a bit more friendly to "outsiders".

Perhaps one idea would be to have a "rust status" website somewhere, where all of the teams/WGs can asynchronously update what they've been up to recently, kind of like the WG checkins that happen at the T-compiler meetings on Thursday. I don't feel that would impose undue overhead, though perhaps others are more qualified to comment.

Reply:
> I would argue that the following are highly desirable (and maybe to some extent, crucial):

Can you articulate why it is crucial?...

> * It should be easy for contributors (especially new contributors) to see what is being worked on, what priorities are there, etc

...my sense is that the reasoning wrt. crucial here is that we need old contributors to stay and new ones to come in for us to get things done (e.g. all the technical debt that we need to work off). That does seem important. The needs of contributors, as compared to users, do however seem different. In particular, I would expect that "what is being worked on" and "what priorities there are" would be clarified, above the feature-gate level, primarily what by project / working groups there are. 

> * It should be easy for users to see where the language is headed, what features are likely or unlikely to land, etc

These two aspects are quite distinct from each other. Specifically, "where the language is headed" is something which requires a very broad overview of a large amount of features (well, except for features which are very large, in which case see discussion below re. tracking issues for feature gates).

Meanwhile, the likelihood of a feature landing or not primarily depends on the number of open bugs, open design questions, and the strength of the test suite. This is not always clear from a tracking issue, especially old ones without `F-my_gate` labels and without a noted implementation history, but that is a matter of technical debt that I'm trying to pay off. Let's take a concrete example of a new-style tracking issue.. Do you find that the information in https://github.com/rust-lang/rust/issues/54883 (which I've linked quite a few times, heh...) makes the current status and stabilization-right-now unclear? If so, how?

> Note that these goals are broader than a single issue; they are project-level, so tracking issues are insufficient to solve this problem. I think this is why there are often questions like "any progress on this?" on tracking issues.

Typically, I find that the language team works most of the time on a granularity level of individual feature gates, PRs, and issues. Sometimes we also have larger efforts, and then those should be tracked using meta-issues (e.g. https://github.com/rust-lang/rust/issues/63066 and https://github.com/rust-lang/rust/issues/57563). So I think a lot of the high level overview information just does not exist because we do not work that way.

I do indeed see people leaving *"any progress on this?"* or *"what's he status?"* on a tracking issue from time to time.
- Sometimes that's because the tracking issue is genuinely messy (https://github.com/rust-lang/rust/issues/31844). This, the `specialization` tracking issue is an example of a poorly moderated old-style issue, which largely does not actually track anything. With new tracking issues I setup things differently: I note explicitly (and enforce) that tracking issues are for tracking implementation progress and status updates. Design questions, bugs, and other larger pieces of discussions are moved to issues or other places. This ensures that the relevant updates are not lost in a see of irrelevant information to sift through and ignore for a user (and for contributors & team members).
- Sometimes it is because the person cannot be bothered to read the OP which details the status that is clearly provided.
- Other times, the person wants really means that there *should* be more progress on an issue whereas there isn't (*"any updates?"* really means *"please make progress"*, but most of the time that's not an actionable comment and thus unhelpful).

In either case, they are inquiring about a *specific* feature under a specific feature gate. So I don't understand how e.g. a full accounting of the implementation history, with updates whenever a PR lands (as in e.g., https://github.com/rust-lang/rust/issues/54883) does not satiate what the person wants to know.

> To be clear, I'm not arguing that we should actually make the giant kanban board. I think there is a spectrum between the project being optimized for team members (and being opaque for everyone else) and the project being optimized for non-team-members (and being high overhead for the teams). I am arguing that perhaps we need adjust where on the spectrum we fall to be a bit more friendly to "outsiders".

The language team has already substantially increased its transparency, with language team minutes being published in the repo, and [all meetings being public and open for everyone to attend](https://github.com/rust-lang/lang-team/#meeting-calendar), as well as [posted to YouTube](https://www.youtube.com/watch?v=kU7is4KeURo). As aforementioned, I've also tried to improve the organization of tracking issues to link PRs, bug reports, and tracking issues together. Just this baseline work itself is time consuming as right now, I have to watch PRs and issues to make sure the labels are added (though I haven't communicated the new system as well as I could have).

> Perhaps one idea would be to have a "rust status" website somewhere, where all of the teams/WGs can asynchronously update what they've been up to recently, kind of like the WG checkins that happen at the T-compiler meetings on Thursday. I don't feel that would impose undue overhead, though perhaps others are more qualified to comment.

If there are more people who can put in the hours do this managerial work, then sure, that can work. However, I am concerned that this will in reality result in an increase of work for the team members themselves, work which is also unevenly balanced between the members by the way (e.g. I think much of this falls to Niko and myself). *If* that is the consequence then I think that is in fact an undue burden. I personally cannot add more hours to the essentially unpaid (more-than-)full time work I'm doing today.

Reply:
I would expect that the additional organization tracking effort would come from triage.  They already have their ear to the ground on all things Rust to begin with.

Naturally, if everything is already working well then triage won't even have to ask extra questions from team/group members because it'll all be well documented in the ways that you say. There would only need to be "hey tell me what's going on lately" questions if the suggested methods don't pan out (which will naturally happen from time to time when people get extra busy).

Of course, I'm also told that they feel a little overwhelmed sometimes, so as always "we need to increase our bandwidth" pops up again. At least with triage it's a lot easier to on-board new members.

Reply:
I agree with @mark-i-m but for I would add a few different, if complementary, reasons. My sense is that it would be great if we can do a better job of exposing what is going on not only for *newcomers* but *for the team members*. I think it's hard to keep up and having more communication about "these are the things we are tracking" would help us in doing so. It would also help us ensure that we're paying attention to the things we mean to and that we're achieving good progress towards our objectives (it's very easy to get lost in reactive work, I think).

In the lang team meetings, for example, I think that the project group focus has helped me to have a sense of what's going on, although I think the current setup isn't ideal (and in particular it often doesn't include things that should still, imo, be tracked, like tracking issues and features that we are moving to stability).

Similarly, one of the motivations behind the compiler-team working groups was the idea "we should absolutely be writing down what is **actively** happening" and making it obvious where to go to ask questions about it. 

I don't really think we've hit the nail on the head here in terms of setup, and as a result, things aren't getting used quite as much as I would like. But I think the goal is useful and important and worthy of further consideration.

On that final point, I definitely agree with @Centril that the "busywork factor" should not be overlooked. I am obviously a big fan of IntoRust and so forth, but I've been negligent at posting some announcement style posts lately, because there's just a lot to keep up with. I suspect there's room for us to innovate a bit on how to handle these things in a lightweight fashion.